Added the sent event logging to SendMsg

varunrisbud opened the pull request:
I am not sure how to test the changes in store/src/java/com/zimbra/cs/service/mail/SendMsg.java.
I want to send an email and see if it logging an event. I guess I can directly call the SendMsg::handle() but not sure how to construct the parameters.

Took a look at SendMsgTest::testSendFromDraft(). I think I will need to do something similar. However, not sure what element/attributes should I set in the request?

varunrisbud committed:
Added the sent event logging to SendMsg

ilyusha commented on store/src/java/com/zimbra/cs/service/mail/SendMsg.java:
```diff
@@ -285,6 +287,22 @@ public static ItemId doSendMessage(OperationContext oc, Mailbox mbox, MimeMessag
+                    EventLogger.getEventLogger().log(singleRecipientEvent);
+                }
+            } catch (MessagingException e) {
+                e.printStackTrace();
```
We generally don't want to just print the stack trace of an error; in this case I would log a warning that the event could not be logged. Since the _event_ logger category will be used for successful events, this could go in the _soap_ or _misc_  categories.

varunrisbud committed:
Changed the implementation of EventLogger to Observer pattern

varunrisbud committed:
Added test case to see that correct events are generated for sent mail

varunrisbud committed:
Added blocking queue for events and using threads to notify about the events to log handlers

ilyusha commented on store/src/java/com/zimbra/cs/event/logger/EventLogger.java:
```diff
@@ -4,49 +4,71 @@
+                        logHandler.log(event);
+                    }
+                } catch (InterruptedException e) {
+                    e.printStackTrace();
```
Maybe we should have two log categories - "events" for regular event-related messages and errors, and "eventlog" for the actual separate log of successful events?

ilyusha commented on store/src/java/com/zimbra/cs/event/logger/EventLogger.java:
```diff
@@ -4,49 +4,71 @@
 
-    private EventLogger() {}
+    private EventLogger() {
+        executorService.submit(new EventNotifier());
```
Won't this only result in one executor thread polling the queue, even though the thread pool can handle 2?

ilyusha commented on store/src/java/com/zimbra/cs/event/logger/EventLogger.java:
```diff
@@ -4,49 +4,71 @@
+    public boolean log(Event event) {
+        try {
+            return eventQueue.add(event);
+        } catch (IllegalStateException e) {
```
Do you know what scenarios would lead to this being thrown from a LinkedBlockingQueue?

ilyusha committed:
new LDAP attrs for event logging

ilyusha committed:
EventLoggerCallback

varunrisbud committed:
Added the shutdown method for the Event Logger

ilyusha committed:
EventLogHandler Factories

varunrisbud committed:
Added the functionality to drain the event queue when shutting down the event logger

ilyusha committed:
EventLogger updates

- EventLogHandler factory support
- Each EventNotifier thread has its own handler instances
- Configuration via ConfigProvider interface
- Other small changes

ilyusha committed:
startup/shutdown event logger in Zimbra class

ilyusha committed:
removed blank line in zimbraIndexURL description that was breaking LDAP

varunrisbud committed:
Added test cases for EventLogger. Fixed minor issues

ilyusha committed:
minor changes to EventLogger

ilyusha committed:
updates to SendMsgTest

ilyusha committed:
allow multiple event handlers for same factory

varunrisbud committed:
Added SENT event convinience methods
